characteristic_notify/device_code.js

```diff
@@ -0,0 +1,91 @@
+/**
+ *  Copyright 2021 Google LLC
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+const data = new Uint32Array([1]);
+let isConnected = false;
+let debug = '?';
+
+function updateScreen() {
+  if (typeof g === 'undefined') {
+    // Device has no graphics device (i.e. screen).
+    return;
+  }
+  g.clear();
+  g.setFontBitmap();
+  g.drawString('Current Notify Value:');
+  g.drawString('Connected: ' + isConnected, 0, g.getHeight() - 10);
+  let msg = 'Dbg: ' + debug;
+  g.drawString(msg, g.getWidth() - g.stringWidth(msg), g.getHeight() - 10);
+  g.setFontVector(40);
+  let val = data[0];
+  g.drawString(val, (g.getWidth() - g.stringWidth(val)) / 2, 12);
+
+  g.flip();
+}
+
+function updateValue() {
+  data[0] = data[0] + 1;
+  if (data[0] == 9999) {
+    // Limit to four digits for display on small screens.
+    data[0] = [0];
+  }
+  if (isConnected) {
+    NRF.updateServices({
+      '02fc549a-244c-11eb-adc1-0242ac120002': {
+        '02fc549a-244c-11eb-adc1-0242ac120002':
+            {value: data.buffer, notify: true}
+      }
+    });
+  }
+  updateScreen();
+}
+
+function onInit() {
+  // Put into a known state.
+  digitalWrite(LED, isConnected);
+
+  NRF.setServices({
+    '02fc549a-244c-11eb-adc1-0242ac120002': {
+      '02fc549a-244c-11eb-adc1-0242ac120002': {
+        value: data.buffer,
+        broadcast: false,
+        readable: true,
+        writable: false,
+        notify: true,
+        description: 'Notify characteristic',
+      }
+    }
+  });
+
+  NRF.on('disconnect', (addr) => {
+    // Provide feedback that device no longer connected.
+    digitalWrite(LED, 0);
+    isConnected = false;
+    debug = addr;
+    updateScreen();
+  });
+
+  NRF.on('connect', (reason) => {
+    digitalWrite(LED, 1);
+    isConnected = true;
+    debug = reason;
+    updateScreen();
+  });
+
+  updateScreen();
+
+  setInterval(updateValue, 2000);
```

Comment:
The next version changes values every 40 msec. on the device. When I tested it at 20 msec. I got no notifications at all. I didn't research this, but I suspect we're hitting some Espruino/device BT limit. I wait for 100 changes, but could increase. 1000 changes @ 40ms would be 40 seconds. I can bump it up a little - LMK.

Reply:
We'll want tests which push the API harder but starting with an "easy" test is good because when it goes wrong it is easier to debug. I'm fine with landing the original version which only waited for 2 updates 1000ms or 2000ms apart.

The tests that push more data will necessarily run longer. Since we've seen developers reporting issues which occur after thousands of updates we will have to have such long running tests but should be able to verify basic functionality quickly.

Reply:
Got it. I'm happy to parameterize this test for that use-case, or come up with some endurance tests for them as well.